docs/adr/0006-server-arch.md

```diff
@@ -0,0 +1,185 @@
+# 0006. Server Architecture
+
+Date: 7th June 2019
+
+## Status
+
+Proposed
+
+## Context
+
+The server-side code has a number of issues with it which make it difficult to
+test and difficult to work with. The cause of this is because in general the
+code does not adhere to the SOLID principles.
+
+This ADR proposes a new way of layering the code to be able to solve these issues.
+
+## New Code Architecture
+
+Below is a diagram and some example of how the new code architecture will work.
+It is imagined that for each domain (Submission, QC, Assign, Review, Decision)
+there will be a package that implements this pattern for that domain.
+
+<pre>
+
+    +------------------------------------------+
+    |            GraphQL Resolvers             |
+    +------------------------------------------+
+               |
+    +----------------------+       +-----------+
+    |      use cases       +<------+  config   |
+    +----------------------+       +-----------+
+        |           |                    |
+        |     +------------+       +-----------+
+        |     | aggregates |       | services  |
+        |     +------------+       +-----------+
+        |           |(n)
+    +------------------------------------------+
+    |                 models                   |
+    +------------------------------------------+
+    |             data access layer            |
+    +------------------------------------------+
+
+</pre>
+
+---
+
+## General Principles
+
+Each of the following sections describes the general principles that are to
+be used when writing code. This approach uses SOLID Principles.
+
+### GraphQL Resolvers
+
+- The external interface uses graphQL and these are handled with the resolvers.
+- The resolvers should contain no logic themselves. Rather this is delegated
+  to a `use-case` making the resolvers a thin wrapper around the controller logic.
+- Rationale is that it should be easy to implement another interface on the same
+  logic.
+
+#### Example
+
+```
+const { getSubmissionUseCase, } = require('use-cases')
+
+const resolvers = {
+  Query: {
+    async manuscript(_, { id }, { user }) {
+      return getSubmissionUseCase(id, user)
+    },
+    ...
+  }
+  ...
+```
+
+### Use-Cases
+
+- In general this layer will have an interface that maps one-to-one with the
+  graphQL layer.
+- For a given use-case it will require certain services and config - these
+  dependencies will be created by this layer and injected (when needed) into
```

Comment:
injection + "when needed": smell to me (as in partially constructed model object, with dependencies left to null or that change often). Perhaps you meant: passed in as parameters to models methods to avoid a dependency going from the lower to the upper layer?

Reply:
Ah the "when needed" was meant to refer to the fact that not all models need the same services (for example `Team` does not need to know about the "Storage" service. I'll clarify this - thanks